internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,395 @@
+	if err != nil {
+		return err
+	}
+	rep := m.config.repetitions()
+	tk := &TestKeys{
+		Domain:      host,
+		Repetitions: rep,
+	}
+	measurement.TestKeys = tk
+
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	// create UDP socket
+	conn, err := m.config.networkLibrary().ListenUDP("udp", &net.UDPAddr{})
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	conn.SetReadDeadline(time.Now().Add(time.Duration(rep*m.config.timeout()) * time.Millisecond))
+
+	sendInfoMap := make(map[string]*sendInfo)
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// this goroutine sends a ping request every second
+	go func() {
+		for i := int64(0); i < rep; i++ {
+			sess.Logger().Infof("PING %s", service)
+
+			sent, dstID, srcID := buildPacket()  // build QUIC Initial packet
+			_, err = conn.WriteTo(sent, udpAddr) // send Initial packet
+			sendTime := time.Now()
+			if err != nil { // if send failed, log that ping attempt here
+				tk.Pings = append(tk.Pings, &SinglePing{
+					Failure:     archival.NewFailure(err),
+					RequestTime: formatTime(sendTime),
+				})
+				continue
+			}
+			m.mu.Lock()
+			sendInfoMap[hex.EncodeToString(srcID)] = &sendInfo{dstID: dstID, srcID: srcID, raw: sent, sendTime: sendTime}
+			m.mu.Unlock()
+			<-ticker.C
+		}
+		wg.Done()
+	}()
+
+	for {
+		resp, respTime, err := m.waitResponse(conn) // wait for server response
+		if err != nil {
+			break // leave loop when the read timeout occured
+		}
+		supportedVersions, dst, err := m.DissectVersionNegotiation(resp) // dissect server response
+		if err != nil {
+			sess.Logger().Infof(fmt.Sprintf("response dissection failed: %s", err))
+			continue
+		}
+		var (
+			req *sendInfo
+			ok  bool
+		)
+		m.mu.Lock()
+		if req, ok = sendInfoMap[dst]; !ok {
+			continue // we have not send a request for this response, so let's discard it for now
+		}
+		delete(sendInfoMap, dst)
+		m.mu.Unlock()
+
+		tk.Pings = append(tk.Pings, &SinglePing{
+			ConnIdDst:         req.dstID,
+			ConnIdSrc:         req.srcID,
+			Failure:           nil,
+			Request:           &model.ArchivalMaybeBinaryData{Value: string(req.raw)},
+			RequestTime:       formatTime(req.sendTime),
+			Response:          &model.ArchivalMaybeBinaryData{Value: string(resp)},
+			ResponseTime:      formatTime(*respTime),
+			SupportedVersions: supportedVersions,
+		})
+		sess.Logger().Infof("PING got response from %s", service)
+
+		if len(tk.Pings) >= int(rep) {
+			break // leave loop when we collected all ping responses
+		}
+	}
+	wg.Wait()
+	// log all ping requests that have not been answered
+	timeoutErr := errors.New("i/o timeout")
+	for _, req := range sendInfoMap {
+		tk.Pings = append(tk.Pings, &SinglePing{
+			ConnIdDst:         req.dstID,
+			ConnIdSrc:         req.srcID,
+			Failure:           archival.NewFailure(timeoutErr),
+			Request:           &model.ArchivalMaybeBinaryData{Value: string(req.raw)},
+			Response:          nil,
+			SupportedVersions: nil,
+		})
+	}
```

Comment:
So, with the changes I recommended above, the main loop becomes something like:

```Go
for {
  select {
  case ping := <-pingch:
    // store the ping info into the map (no mutex needed since the map is owned by Run)
  case pong := <-pongch:
    // figure out if we have a corresponding ping (again no mutex needed)
  case <-ctx.Done():
    break
  }
}
```